src/metabase/lib/equality.cljc

```diff
@@ -104,106 +109,365 @@
+                             (-> (resolve-field-id query stage-number field-id)
+                                 (dissoc :id ; Remove any :id to force the ref to use the field name.
+                                         :lib/desired-column-alias) ; Hack: resolving by name doesn't work if this is present
+                                 lib.ref/ref))))]
+     (if converted
+       ;; If any of the needles were not found, and were converted successfully, try to match them again.
+       (merge matches (find-closest-matches-for-refs converted haystack opts))
+       ;; If we found them all, just return the matches.
+       matches))))
+
+(defn find-closest-matching-ref
+  "Given a target `a-ref` and a list `refs-or-cols` of `MetadataColumns` or refs, and finds the closest match.
+  Returns the value from `refs-or-cols` which most closely corresponds to `a-ref`, or nil if nothing matches.
+
+  See [[find-closest-matches-for-refs]] for details of how the approximate matching works. (Where
+  [[find-closest-matches-for-refs]] would return `{column 0}`, this returns just `column`.)"
+  ([a-ref refs-or-cols]
+   (find-closest-matching-ref a-ref refs-or-cols {}))
+  ([a-ref refs-or-cols opts]
+   (->> (find-closest-matches-for-refs [a-ref] refs-or-cols opts)
+        keys
+        first))
+
+  ([query stage-number a-ref refs-or-cols]
+   (find-closest-matching-ref query stage-number a-ref refs-or-cols {}))
+  ([query stage-number a-ref refs-or-cols opts]
+   (->> (find-closest-matches-for-refs query stage-number [a-ref] refs-or-cols opts)
+        keys
+        first)))
+
+(defn- annotate-index [xs]
+  (map-indexed (fn [i x]
+                 (when x
+                   (vary-meta x assoc ::index i)))
+               xs))
+
+(defn- cascading-matches
+  "For each of `ref-fns`, this does the following:
+  - Map `(first ref-fns)` over the `haystack-metadatas`.
+  - Annotate these results with metadata to map them back to the original `haystack-metadatas`.
+  - Call [[find-closest-matches-for-refs]] with `needles` and this list of refs.
+  - Map the results back to use the original `haystack-metadatas` as keys, rather than the refs.
+  - Merge the results right-to-left, so that **earlier matches win**.
+  - Remove any matched `needles` and `haystack-metadatas` from the inputs.
+  - If both inputs are nonempty, continue to the next `ref-fns`, if any."
+  [needles haystack-metadatas opts ref-fns]
+  (let [original-needles (annotate-index needles)]
+    (loop [current-needles         original-needles
+           current-haystack        haystack-metadatas
+           [ref-fn & more-ref-fns] ref-fns
+           result                  {}]
+      (cond
+        (or (empty? current-needles)
+            (empty? current-haystack)
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
+        :else
+        (let [haystack-ref->col  (m/index-by ref-fn current-haystack)
```

Comment:
Should it be an error if `(< (count haystack-ref->col) (count current-haystack))`?